packages/angular/cli/utilities/bep.ts

```diff
@@ -0,0 +1,54 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import * as fs from 'fs';
+
+
+export interface BuildEventMessage {
+  id: {};
+  [key: string]: {};
+}
+
+export class BepGenerator {
+  private constructor() {}
+
+  static createBuildStarted(command: string, time?: number): BuildEventMessage {
+    return {
+      id: { started: {} },
+      started: {
+        command,
+        start_time_millis: time == undefined ? Date.now() : time,
+      },
+    };
+  }
+
+  static createBuildFinished(code: number, time?: number): BuildEventMessage {
+    return {
+      id: { finished: {} },
+      finished: {
+        finish_time_millis: time == undefined ? Date.now() : time,
+        exit_code: { code },
+      },
+    };
+  }
+}
+
+export function appendEvent(
+  path: string,
+  event: BuildEventMessage,
+): void {
+  const exists = fs.existsSync(path);
+
+  let content = '';
+  if (exists) {
+    content = fs.readFileSync(path, 'utf8');
```

Comment:
this seems wasteful, can't we just open the file in append mode rather than read its contents?
Even better, could we hold the filehandle open and just write/flush to the buffer?

Reply:
yes. it is.  However, currently the frequency of writes is quite minimal.  I was going to eventually create a `BepJsonWriter` class that encapsulated the logic as you described along with the message generation helpers.

Reply:
`BepJsonWriter` introduced to better encapsulate the message writing.